Mathlib/RingTheory/Flat/CategoryTheory.lean

```diff
@@ -0,0 +1,193 @@
+  right tensoring with `M` preserves finite limits, i.e. the functor `M ⊗ -` is left exact.
+
+- `Module.Flat.iff_lTensor_preserves_shortComplex_exact`: an `R`-module `M` is flat if and only if
+  for every short exact sequence `0 ⟶ A ⟶ B ⟶ C ⟶ 0`, `0 ⟶ M ⊗ A ⟶ M ⊗ B ⟶ M ⊗ C ⟶ 0` is also
+  a short exact sequence.
+
+- `Module.Flat.iff_rTensor_preserves_shortComplex_exact`: an `R`-module `M` is flat if and only if
+  for every short exact sequence `0 ⟶ A ⟶ B ⟶ C ⟶ 0`, `0 ⟶ A ⊗ M ⟶ B ⊗ M ⟶ C ⊗ M ⟶ 0` is also
+  a short exact sequence.
+
+- `Module.Flat.higherTorIsoZero`: if an `R`-module `M` is flat, then `Torⁿ(M, N) ≅ 0` for all `N`
+  and all `n ≥ 1`.
+
+## TODO
+
+- Prove that vanishing `Tor`-groups implies flat.
+
+-/
+
+universe u
+
+open CategoryTheory MonoidalCategory Abelian
+
+namespace Module.Flat
+
+variable {R : Type u} [CommRing R] (M : ModuleCat.{u} R)
+
+lemma lTensor_shortComplex_exact [Flat R M] (C : ShortComplex $ ModuleCat R) (hC : C.Exact) :
+    C.map (tensorLeft M) |>.Exact := by
+  rw [ModuleCat.iff_shortComplex_exact, Eq.comm, ← LinearMap.exact_iff]
+  exact lTensor_exact M $ LinearMap.exact_iff.2 $ Eq.symm $
+    ModuleCat.iff_shortComplex_exact _ _ C.zero |>.1 hC
+
+lemma rTensor_shortComplex_exact [Flat R M] (C : ShortComplex $ ModuleCat R) (hC : C.Exact) :
+    C.map (tensorRight M) |>.Exact := by
+  rw [ModuleCat.iff_shortComplex_exact, Eq.comm, ← LinearMap.exact_iff]
+  exact rTensor_exact M $ LinearMap.exact_iff.2 $ Eq.symm $
+    ModuleCat.iff_shortComplex_exact _ _ C.zero |>.1 hC
+
+lemma iff_lTensor_preserves_shortComplex_exact :
+    Flat R M ↔
+    ∀ (C : ShortComplex $ ModuleCat R) (_ : C.Exact), (C.map (tensorLeft M) |>.Exact) :=
+  ⟨fun _ _=> lTensor_shortComplex_exact _ _, fun H => by
+    rw [iff_lTensor_exact]
+    intro N N' N'' _ _ _ _ _ _ f g h
+    specialize H (.mk (ModuleCat.ofHom f) (ModuleCat.ofHom g)
+      (DFunLike.ext _ _ h.apply_apply_eq_zero)) (ModuleCat.shortComplex_exact _ _ $ Eq.symm $
+        LinearMap.exact_iff |>.1 $ h)
+    rw [LinearMap.exact_iff, Eq.comm]
+    rw [ModuleCat.iff_shortComplex_exact] at H
+    convert H⟩
+
+lemma iff_rTensor_preserves_shortComplex_exact :
+    Flat R M ↔
+    ∀ (C : ShortComplex $ ModuleCat R) (_ : C.Exact), (C.map (tensorRight M) |>.Exact) :=
+  ⟨fun _ _=> rTensor_shortComplex_exact _ _, fun H => by
+    rw [iff_rTensor_exact]
+    intro N N' N'' _ _ _ _ _ _ f g h
+    specialize H (.mk (ModuleCat.ofHom f) (ModuleCat.ofHom g)
+      (DFunLike.ext _ _ h.apply_apply_eq_zero)) (ModuleCat.shortComplex_exact _ _ $ Eq.symm $
+        LinearMap.exact_iff |>.1 $ h)
+    rw [LinearMap.exact_iff, Eq.comm]
+    rw [ModuleCat.iff_shortComplex_exact] at H
+    convert H⟩
+
+open scoped MonoidalCategory in
+set_option maxHeartbeats 400000 in
+noncomputable instance [flat : Flat R M] {X Y : ModuleCat.{u} R} (f : X ⟶ Y) :
+    Limits.PreservesLimit (Limits.parallelPair f 0) (tensorLeft M) where
```

Comment:
After you have obtained `iff_lTensor_preserves_shortComplex_exact`, I believe that the rest should be obtained from a (mission) API about additive functors between abelian categories: their preserves exactness iff they preserves both finite limits and colimits. Then, I think we should develop this API for general characterizations of left exact/right exact/exact functors between abelian categories in terms of their behaviour on exact short complexes.

Reply:
A suggestion is to keep in this PR only the beginning of the file until `iff_rTensor_preserves_shortComplex_exact`. Then, depending on how you feel about the general API I am suggesting above, you may do it, or I may do it if you prefer.

Reply:
I will try to open a PR about this suggestion this week!

Reply:
I have removed everything else except `iff_rTensor_preserves_shortComplex_exact ` and above. I think I will wait for #14520 to golf the proof a bit more

Reply:
> A suggestion is to keep in this PR only the beginning of the file until `iff_rTensor_preserves_shortComplex_exact`. Then, depending on how you feel about the general API I am suggesting above, you may do it, or I may do it if you prefer.

I would want to use this as an opportunity to learn more about the homological part of the library, but my code would probably not be optimal. If you think it would be more efficient to open the PR yourself, I wouldn't mind either.

Reply:
Please do it!